Feature/vfd swmr - adding the log feature

kyang2014 opened the pull request:
1. Add VFD SWMR log reporting function and macro to facilitate the generation of debugging info in the VFD SWMR.

The setup and close the log file are also implemented. 
The code under /src are generally added with comments  that include 'Kent' for easy to review.

2. I also add three testing tags" FILE_OPEN", "FILE_CLOSE" and "EOT_PROCESSING_TIME".
3. The testing program is vfd_swmr_log_writer under /test. Read the description of the source code or run usage will show how to generate a log file.

kyang2014 committed:
Test to just write a log file for H5Fopen. Add a testing routine vfd_swmr_log_writer.c.

kyang2014 committed:
Clean up H5Fint.c, ready to implement the report function.

github-actions[bot] committed:
Committing clang-format changes

kyang2014 committed:
skeleton of the logging report function

kyang2014 committed:
Merge branch 'vfd_swmr_log' of https://github.com/kyang2014/hdf5 into vfd_swmr_log

github-actions[bot] committed:
Committing clang-format changes

kyang2014 committed:
Add the log entry report function, also add logs for 'File open','File close' and 'EOT processing time'

github-actions[bot] committed:
Committing clang-format changes

kyang2014 committed:
Make sure to check the file point.

kyang2014 committed:
Merge branch 'vfd_swmr_log' of https://github.com/kyang2014/hdf5 into vfd_swmr_log

github-actions[bot] committed:
Committing clang-format changes

kyang2014 committed:
Revise the H5Fopen log test.

kyang2014 committed:
Merge branch 'vfd_swmr_log' of https://github.com/kyang2014/hdf5 into vfd_swmr_log

github-actions[bot] committed:
Committing clang-format changes

kyang2014 committed:
Still need to support the NULL pointer in the log report function.

kyang2014 committed:
add macro, need to debug an error caused by using the macro

github-actions[bot] committed:
Committing clang-format changes

kyang2014 committed:
Update comments, formats etc.

github-actions[bot] committed:
Committing clang-format changes

kyang2014 committed:
Remove the compiler warnings.

github-actions[bot] committed:
Committing clang-format changes

kyang2014 committed:
Add comments, descriptions for the VFD log feature.

github-actions[bot] committed:
Committing clang-format changes

kyang2014 committed:
Modify the description a bit.

kyang2014 committed:
Merge branch 'vfd_swmr_log' into feature/vfd_swmr

kyang2014 committed:
Add vfd_swmr_log_writer.c to the MANIFEST.

kyang2014 committed:
Modify comments.

github-actions[bot] committed:
Committing clang-format changes

kyang2014 committed:
Revise a comment to trigger another github check

kyang2014 committed:
Add macros to make windows ignore the HDgetclock_time().

github-actions[bot] committed:
Committing clang-format changes

kyang2014 committed:
Use HDF5 timer function

github-actions[bot] committed:
Committing clang-format changes

kyang2014 committed:
Update comments to use the HDF5 timer for the log feature.

github-actions[bot] committed:
Committing clang-format changes

vchoi-hdfgroup commented on src/H5Fint.c:
```diff
@@ -2013,7 +2013,25 @@ H5F_open(const char *name, unsigned flags, hid_t fcpl_id, hid_t fapl_id)
+
+    /* Set up the VFD SWMR LOG file */
+    /* Kent*/
+    if (vfd_swmr_config_ptr->version) {
```
See if you can put this section of code to line #2112 so that VFD SWMR changes will be together in one place.

  /* Checked if configured for VFD SWMR */
    if (H5F_VFD_SWMR_CONFIG(file)) {
    ........

kyang2014 replied:
Will do.

kyang2014 replied:
Done

vchoi-hdfgroup commented on src/H5Fpkg.h:
```diff
@@ -608,4 +618,52 @@ H5_DLL htri_t H5F__same_file_test(hid_t file_id1, hid_t file_id2);
+        }                                                                                                    \
+    } while (0)
+
+#define H5F_POST_VFD_SWMR_LOG_ENTRY_RELEASE(fp, entry_type_code, max_code, log_info)                         \
```
I suggest to change the name to H5F_POST_VFD_SWMR_LOG_ENTRY_PRODUCTION because H5F_POST_VFD_SWMR_LOG_ENTRY_RELEASE initially gives me the impression of releasing resources for data structures.

kyang2014 replied:
Will do.

kyang2014 replied:
Done

vchoi-hdfgroup commented on test/vfd_swmr_log_writer.c:
```diff
@@ -0,0 +1,2929 @@
+ * help@hdfgroup.org.                                                        *
+ * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
+
+/* Description of this program:
```
Does this test do anything in particular to verify the logging feature?

Also, if this test is the same as the group performance test, instead of adding this new test, maybe add an option to the group performance test to enable the logging feature?

kyang2014 replied:
I like this idea as long as you and John are comfortable if I put this into the performance test. Please verify so I can merge this test to the group performance test.

vchoi-hdfgroup replied:
I am ok with that.  Maybe check with John also.

kyang2014 replied:
Sent an email to John. Haven't received his reply. Think this should be OK to him. Go ahead to merge it.

vchoi-hdfgroup commented on src/H5Fvfd_swmr.c:
```diff
@@ -319,7 +350,20 @@ H5F_vfd_swmr_close_or_flush(H5F_t *f, hbool_t closing)
+     * Please REVIEW to ensure this is the right place to
+     * close the log file.
+     */
+    if (shared->vfd_swmr_log_on) {
```
Do we close the log file if it is flushing not file closing?

kyang2014 replied:
I don't know if this close the log file when the file is flushing.  As in the comments, I don't know if this is the right place to close the log file. If you know there is a place that I can close the log file, please let me know. I checked H5Fclose first but there is no H5F_t in that routine. So this is the best place I can find. If you are not sure, I suggest we keep here until you and John can find the right place to close the log file.

kyang2014 replied:
Vailin, I think I find the solution. 
I can just add a check if this is for file closing:
Like this
 ```
if (shared->vfd_swmr_log_on && closing) {
       H5_timer_stop(&(shared->vfd_swmr_log_start_time));
        HDfclose(shared->vfd_swmr_log_file_ptr);
} 
```
Then the log file will only be closed when this function is handling the close of the file.

vchoi-hdfgroup replied:
ok

vchoi-hdfgroup commented on src/H5FDvfd_swmr_private.h:
```diff
@@ -84,4 +84,8 @@ H5_DLL herr_t H5F_vfd_swmr_insert_entry_eot(struct H5F_t *f);
 H5_DLL void   H5F_vfd_swmr_update_entry_eot(eot_queue_entry_t *);
 H5_DLL herr_t H5F_dump_eot_queue(void);
 
+/***************************************/
```
Remove these lines if nothing is added?

kyang2014 replied:
OK

vchoi-hdfgroup commented on src/H5Fvfd_swmr.c:
```diff
@@ -883,6 +943,18 @@ H5F_vfd_swmr_writer_end_of_tick(H5F_t *f, hbool_t wait_for_reader)
+        if (H5_timer_get_times(shared->vfd_swmr_log_start_time, &current_time) < 0)
+            HGOTO_ERROR(H5E_FILE, H5E_CANTGET, FAIL, "can't get time from H5_timer_get_times")
+        end_elapsed_time = current_time.elapsed;
+        log_msg          = HDmalloc(48);
```
Error check for HDmalloc(48)?
Use define for 48 instead of a constant?

kyang2014 replied:
Vailin, I can define a constant. I didn't do this because this part of code is only valid when the logging is turned on. Also this is an example. I am not sure if we want to keep this.

vchoi-hdfgroup replied:
ok

vchoi-hdfgroup commented on src/H5Fvfd_swmr.c:
```diff
@@ -1886,3 +1958,46 @@ H5F_vfd_swmr_process_eot_queue(hbool_t entering_api)
+       else obtain the elapsed time in seconds since the log file
+            was created and write the time to the log file. */
+    if (H5_timer_get_times(f->shared->vfd_swmr_log_start_time, &current_time) < 0) {
+        gettime_error = HDmalloc(14);
```
Error check for HDmalloc(14)?
Use define for 14 instead of a constant?

kyang2014 replied:
Same reason. I don't know how HDF5 library is usually handled. If defining a constant is fine even if the logging is not used, I would like to do that.

vchoi-hdfgroup replied:
ok

kyang2014 committed:
1. Close the log file when the file closing flag is true in H5F_vfd_swmr_close_or_flush().
2. Move the log test to the group performance test code with an option.
3. Use constant variables for log message lengths.
4. Misc. clean-up.

github-actions[bot] committed:
Committing clang-format changes

jrmainzer commented on src/H5Fint.c:
```diff
@@ -2210,6 +2227,9 @@ H5F_open(const char *name, unsigned flags, hid_t fcpl_id, hid_t fapl_id)
     ret_value = file;
 
+#if 1 /*Kent: write to the log file when H5F_open ends. Tested, can be commented out if necessary.*/
+    H5F_POST_VFD_SWMR_LOG_ENTRY(file, 1, "File open ends");
```
We should have #defines for the entry_type_code parameters, not just integer constants.

Suggest you put the definitions of these #defines in the same file as your macro wrappers for the logging function proper.

jrmainzer replied:
On further review, probably best to put them adjacent to your declaration of H5Fvfd_swmr_log_tags.

kyang2014 replied:
Done.

kyang2014 committed:
Make entry log code as a #define macro as John suggests.

github-actions[bot] committed:
Committing clang-format changes
